internal/s3manager/s3manager.go

```diff
@@ -0,0 +1,203 @@
+// Package s3manager implements the various functions needed to store and retrieve
+// files from an S3 API compatible endpoint (AWS S3, Minio, etc)
+package s3manager
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/epinio/epinio/helpers/kubernetes"
+	"github.com/google/uuid"
+	minio "github.com/minio/minio-go/v7"
+	"github.com/minio/minio-go/v7/pkg/credentials"
+	"github.com/pkg/errors"
+	"gopkg.in/ini.v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type Manager struct {
+	minioClient       *minio.Client
+	connectionDetails ConnectionDetails
+}
+
+type ConnectionDetails struct {
+	Endpoint string
+	// UseSSL Toggles the use of SSL for the s3 connection. If we deploy
+	// our own thing, linkderd will take care of this and we set useSSL to
+	// false.
+	// External S3 implementations should use SSL.
+	UseSSL          bool
+	AccessKeyID     string
+	SecretAccessKey string
+	Bucket          string
+	Location        string
+}
+
+func NewConnectionDetails(endpoint, key, secret, bucket, location string, useSSL bool) *ConnectionDetails {
+	return &ConnectionDetails{
+		Endpoint:        endpoint,
+		UseSSL:          useSSL,
+		AccessKeyID:     key,
+		SecretAccessKey: secret,
+		Bucket:          bucket,
+		Location:        location,
+	}
+}
+
+// Validate makes sure the S3 settings provided are valid
+// The user should either provide all the mandatory settings
+func (details *ConnectionDetails) Validate() error {
+	allMandatorySet := details.Endpoint != "" &&
+		details.AccessKeyID != "" &&
+		details.SecretAccessKey != "" &&
+		details.Bucket != ""
+	allMandatoryEmpty := details.Endpoint == "" &&
+		details.AccessKeyID == "" &&
+		details.SecretAccessKey == "" &&
+		details.Bucket == ""
+	optionalSet := details.Location != ""
+
+	// If mandatory fields are partly set
+	partlyMandatory := !(allMandatorySet || allMandatoryEmpty)
+	if partlyMandatory {
+		return errors.New("when specifying an external s3 server, you  must set all  mandatory S3 options")
+	}
+
+	// If only optional fields are set
+	if allMandatoryEmpty && optionalSet {
+		return errors.New("do not specify options if using the internal S3 storage")
+	}
+
+	// Either all empty or at least the mandatory fields set - valid.
+	return nil
+}
+
+// New returns an instance of an s3 manager
+func New(connectionDetails ConnectionDetails) (*Manager, error) {
+	useSSL := connectionDetails.UseSSL
+
+	// TODO add support for skip verify and private certs
+	// t, _ := minio.DefaultTransport(useSSL)
+	// t.TLSClientConfig.InsecureSkipVerify = true
```

Comment:
This should only be needed if someone uses and external, yet insecure S3. This should be a different feature, if anyone ever asks for it. I'll remove these for now.